tiatoolbox/models/controller/semantic_segmentor.py

```diff
@@ -0,0 +1,1017 @@
+            stride_shape=ioconfig.stride_shape,
+        )
+        return patch_inputs, patch_outputs
+
+    @staticmethod
+    def filter_coordinates(
+        mask_reader: VirtualWSIReader,
+        bounds: np.ndarray,
+        resolution: Union[float, int] = None,
+        units: str = None,
+    ):
+        """
+        Indicates which coordinate is valid basing on the mask.
+
+        To use your own approaches, either subclass to overwrite or directly
+        assign your own function to this name. In either cases, the function must
+        obey the API defined here.
+
+        Args:
+            mask_reader (:class:`.VirtualReader`): a virtual pyramidal
+              reader of the mask related to the WSI from which we want
+              to extract the patches.
+            bounds (ndarray and np.int32): Coordinates to be checked
+              via the `func`. They must be in the same resolution as requested
+              `resolution` and `units`. The shape of `coordinatess` is (N, K)
+              where N is the number of coordinate sets and K is either 2 for
+              centroids or 4 for bounding boxes. When using the default
+              `func=None`, K should be 4, as we expect the `coordinatess` to be
+              refer to bounding boxes in `[start_x, start_y, end_x, end_y]` format.
+
+        Returns:
+            ndarray: list of flags to indicate which coordinate is valid.
+
+        Examples:
+            >>> # API of function expected to overwrite `filter_coordinates`
+            >>> def func(reader, bounds, resolution, units):
+            ...   # as example, only select first bound
+            ...   return np.array([1, 0])
+            >>> coords = [[0, 0, 256, 256], [128, 128, 384, 384]]
+            >>> segmentor = SemanticSegmentor(model='unet')
+            >>> segmentor.filter_coordinates = func
+
+        """
+        if not isinstance(mask_reader, VirtualWSIReader):
+            raise ValueError("`mask_reader` should be VirtualWSIReader.")
+        if not isinstance(bounds, np.ndarray) or not np.issubdtype(
+            bounds.dtype, np.integer
+        ):
+            raise ValueError("`coordinatess` should be ndarray of integer type.")
+
+        mask_real_shape = mask_reader.img.shape[:2]
+        mask_resolution_shape = mask_reader.slide_dimensions(
+            resolution=resolution, units=units
+        )[::-1]
+        mask_real_shape = np.array(mask_real_shape)
+        mask_resolution_shape = np.array(mask_resolution_shape)
+        scale_factor = mask_real_shape / mask_resolution_shape
+        scale_factor = scale_factor[0]  # what if ratio x != y
+
+        def sel_func(coord: np.ndarray):
+            """Accept coord as long as its box contains bits of mask."""
+            coord_in_real_mask = np.ceil(scale_factor * coord).astype(np.int32)
+            tl_x, tl_y, br_x, br_y = coord_in_real_mask
+            roi = mask_reader.img[tl_y:br_y, tl_x:br_x]
```

Comment:
You could also call these (top, left, bottom, right) or (start_x, start_y, end_x, end_y) as that is what they are called else where in the toolbox. These names are ok but just trying to avoid use using too many names for the same thing.